File: algorithms/php/1448.CountGoodNodesinBinaryTree.php
<?php

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function goodNodes($root) {
        return $this->dfs($root);
    }

    function dfs($node, $path = []) {
        if (empty($path)) {
            $path[] = $node->val;
            return 1 + $this->dfs($node->left, $path) + $this->dfs($node->right, $path);
        }

        if ($node) {
            $cnt = 0;
            if ($this->array_test($path, $node->val)) {
                $cnt = 1;
                echo $node->val;
            }

            $path[] = $node->val;

            return $cnt + $this->dfs($node->left, $path) + $this->dfs($node->right, $path);
        }
    }

    function array_test($array, $value) {
        return !in_array(false, array_map(function($item) use ($value) {
            return $item <= $value;
        }, $array));
    }
}

File: algorithms/cpp/1448.CountGoodNodesinBinaryTree.cpp
#include "functions.h"
#include "TreeNode.h"

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode *root) {
        return dfs(root);
    }

    int dfs(TreeNode *node, int max = INT32_MIN) {
        int good = 0;

        if (node == nullptr) {
            return 0;
        }

        if (node->val >= max) {
            good++;
            max = node->val;
        }

        return good + dfs(node->left, max) + dfs(node->right, max);
    }
};

int main() {
    vector<int> in1{3,3,-1,4,2};
    TreeNode *t1 = make_tree(in1);
    print_tree(t1);
    Solution s;
    cout << s.goodNodes(t1);

    remove_tree(t1);
}